database/seeders/PlayerAchievementsSeeder.php

```diff
@@ -0,0 +1,154 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Database\Seeders;
+
+use App\Models\Game;
+use App\Models\PlayerGame;
+use App\Models\Role;
+use App\Models\User;
+use App\Platform\Actions\ResumePlayerSessionAction;
+use App\Platform\Actions\UpdateDeveloperContributionYieldAction;
+use App\Platform\Actions\UpdateGameMetricsAction;
+use App\Platform\Actions\UpdatePlayerGameMetricsAction;
+use App\Platform\Actions\UpdatePlayerMetricsAction;
+use App\Platform\Enums\AchievementType;
+use Faker\Factory as Faker;
+use Illuminate\Database\Seeder;
+use Illuminate\Support\Carbon;
+
+class PlayerAchievementsSeeder extends Seeder
+{
+    public function run(): void
+    {
+        $maxPlayers = (int) floor(User::count() / 2);
+        $faker = Faker::create();
+
+        Game::where('achievements_published', '>', 0)->each(function (Game $game) use ($maxPlayers, $faker) {
+            $numWinConditions = $game->achievements()->published()->where('type', AchievementType::WinCondition)->count();
+            $numAchievements = $game->achievements()->published()->count();
+
+            $updatePlayerGameMetricsAction = new UpdatePlayerGameMetricsAction();
+            $resumePlayerSessionAction = new ResumePlayerSessionAction();
+
+            $numPlayers = (int) sqrt(random_int(1, $maxPlayers * $maxPlayers));
+            foreach (User::inRandomOrder()->limit($numPlayers)->get() as $user) {
+                $achievementsRemaining = $numAchievements;
+                if ($user->RAPoints + $user->RASoftcorePoints === 0) {
+                    $hardcore = (random_int(0, 1) === 1);
+                } else {
+                    $hardcore = ($user->RAPoints > $user->RASoftcorePoints);
+                }
+                $keepPlayingChance = random_int(75, 100);
+                $num_sessions = 1;
+
+                $date = Carbon::now()->subMinutes(random_int(1000, 1000000)); // between 16 hours ago and ~2 years ago
```

Comment:
More expressive and can even allow for removing the comment:
```php
$date = fake()->dateTimeBetween('-2 years', '-1 hour');
```

Reply:
Faker returns a DateTime, and it seems like there's no easy way to convert that to a Carbon instance. This is what I settled on after several suggestions from Google:
```
$date = Carbon::parse($faker->dateTimeBetween('-3 years', '-2 hours')->format(\DateTime::ATOM));
```